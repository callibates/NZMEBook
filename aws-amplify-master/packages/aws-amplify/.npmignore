__mocks__/**
__tests__/**
coverage/**
docs/**
node_modules/**
src/**
.vscode/**
.DS_Store
*.log
prepend-license.js
prettier.config.json
tsconfig.json
tsfmt.json
tslint.json
typeDoc.js
webpack.config.js