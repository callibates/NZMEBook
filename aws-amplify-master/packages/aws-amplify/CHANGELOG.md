# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="0.4.9-unstable.1"></a>
## [0.4.9-unstable.1](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.9-unstable.0...aws-amplify@0.4.9-unstable.1) (2018-07-03)




**Note:** Version bump only for package aws-amplify

<a name="0.4.9-unstable.0"></a>
## [0.4.9-unstable.0](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8...aws-amplify@0.4.9-unstable.0) (2018-07-02)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8"></a>
## [0.4.8](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.5...aws-amplify@0.4.8) (2018-06-29)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.5"></a>
## [0.4.8-unstable.5](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.4...aws-amplify@0.4.8-unstable.5) (2018-06-29)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.4"></a>
## [0.4.8-unstable.4](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.3...aws-amplify@0.4.8-unstable.4) (2018-06-29)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.3"></a>
## [0.4.8-unstable.3](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.2...aws-amplify@0.4.8-unstable.3) (2018-06-28)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.2"></a>
## [0.4.8-unstable.2](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.1...aws-amplify@0.4.8-unstable.2) (2018-06-27)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.1"></a>
## [0.4.8-unstable.1](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.8-unstable.0...aws-amplify@0.4.8-unstable.1) (2018-06-27)




**Note:** Version bump only for package aws-amplify

<a name="0.4.8-unstable.0"></a>
## [0.4.8-unstable.0](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.7-unstable.2...aws-amplify@0.4.8-unstable.0) (2018-06-27)


### Bug Fixes

* **interactions:** check fulfilled state to trigger onComplete callback ([#1122](https://github.com/aws/aws-amplify/issues/1122)) ([c24b1f1](https://github.com/aws/aws-amplify/commit/c24b1f1))




<a name="0.4.7"></a>
## [0.4.7](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.7-unstable.2...aws-amplify@0.4.7) (2018-06-27)


### Bug Fixes

* **interactions:** check fulfilled state to trigger onComplete callback ([#1122](https://github.com/aws/aws-amplify/issues/1122)) ([c24b1f1](https://github.com/aws/aws-amplify/commit/c24b1f1))




<a name="0.4.7-unstable.2"></a>
## [0.4.7-unstable.2](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.7-unstable.1...aws-amplify@0.4.7-unstable.2) (2018-06-26)


### Bug Fixes

* **integration tests:** CircleCI workflows and Cypress integration testing ([#1071](https://github.com/aws/aws-amplify/issues/1071)) ([bfa4776](https://github.com/aws/aws-amplify/commit/bfa4776))




<a name="0.4.7-unstable.1"></a>
## [0.4.7-unstable.1](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.7-unstable.0...aws-amplify@0.4.7-unstable.1) (2018-06-22)




**Note:** Version bump only for package aws-amplify

<a name="0.4.7-unstable.0"></a>
## [0.4.7-unstable.0](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.6...aws-amplify@0.4.7-unstable.0) (2018-06-22)




**Note:** Version bump only for package aws-amplify

<a name="0.4.6"></a>
## [0.4.6](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.5-unstable.3...aws-amplify@0.4.6) (2018-06-21)




**Note:** Version bump only for package aws-amplify

<a name="0.4.5"></a>
## [0.4.5](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.47...aws-amplify@0.4.5) (2018-06-20)
<a name="0.4.5-unstable.3"></a>
## [0.4.5-unstable.3](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.5-unstable.2...aws-amplify@0.4.5-unstable.3) (2018-06-21)




**Note:** Version bump only for package aws-amplify

<a name="0.4.5-unstable.2"></a>
## [0.4.5-unstable.2](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.5-unstable.1...aws-amplify@0.4.5-unstable.2) (2018-06-21)




**Note:** Version bump only for package aws-amplify

<a name="0.4.5-unstable.1"></a>
## [0.4.5-unstable.1](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.5-unstable.0...aws-amplify@0.4.5-unstable.1) (2018-06-20)




**Note:** Version bump only for package aws-amplify

<a name="0.4.5-unstable.0"></a>
## [0.4.5-unstable.0](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.47...aws-amplify@0.4.5-unstable.0) (2018-06-20)


### Bug Fixes

* **aws-amplify:** Fix missing ServiceWorker export ([46069e2](https://github.com/aws/aws-amplify/commit/46069e2))
* **pushnotification:** revert change in pr 952 ([b8d167c](https://github.com/aws/aws-amplify/commit/b8d167c))
* **pushnotification:** revert change in pr 952 ([257fc40](https://github.com/aws/aws-amplify/commit/257fc40))




<a name="0.4.4"></a>
## [0.4.4](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.33...aws-amplify@0.4.4) (2018-06-04)
<a name="0.4.2-unstable.47"></a>
## [0.4.2-unstable.47](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.46...aws-amplify@0.4.2-unstable.47) (2018-06-19)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.46"></a>
## [0.4.2-unstable.46](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.45...aws-amplify@0.4.2-unstable.46) (2018-06-18)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.45"></a>
## [0.4.2-unstable.45](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.44...aws-amplify@0.4.2-unstable.45) (2018-06-18)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.44"></a>
## [0.4.2-unstable.44](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.43...aws-amplify@0.4.2-unstable.44) (2018-06-16)


### Features

* **interactions:** Interactions category ([#1042](https://github.com/aws/aws-amplify/issues/1042)) ([befb336](https://github.com/aws/aws-amplify/commit/befb336))




<a name="0.4.2-unstable.43"></a>
## [0.4.2-unstable.43](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.42...aws-amplify@0.4.2-unstable.43) (2018-06-13)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.42"></a>
## [0.4.2-unstable.42](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.41...aws-amplify@0.4.2-unstable.42) (2018-06-13)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.41"></a>
## [0.4.2-unstable.41](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.40...aws-amplify@0.4.2-unstable.41) (2018-06-12)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.40"></a>
## [0.4.2-unstable.40](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.39...aws-amplify@0.4.2-unstable.40) (2018-06-11)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.39"></a>
## [0.4.2-unstable.39](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.38...aws-amplify@0.4.2-unstable.39) (2018-06-08)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.38"></a>
## [0.4.2-unstable.38](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.37...aws-amplify@0.4.2-unstable.38) (2018-06-08)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.37"></a>
## [0.4.2-unstable.37](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.36...aws-amplify@0.4.2-unstable.37) (2018-06-07)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.36"></a>
## [0.4.2-unstable.36](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.35...aws-amplify@0.4.2-unstable.36) (2018-06-06)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.35"></a>
## [0.4.2-unstable.35](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.34...aws-amplify@0.4.2-unstable.35) (2018-06-05)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.34"></a>
## [0.4.2-unstable.34](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.33...aws-amplify@0.4.2-unstable.34) (2018-06-04)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.33"></a>
## [0.4.2-unstable.33](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.32...aws-amplify@0.4.2-unstable.33) (2018-06-04)


### Bug Fixes

* **aws-amplify:** Fix missing ServiceWorker export ([46069e2](https://github.com/aws/aws-amplify/commit/46069e2))
* **pushnotification:** revert change in pr 952 ([b8d167c](https://github.com/aws/aws-amplify/commit/b8d167c))
* **pushnotification:** revert change in pr 952 ([257fc40](https://github.com/aws/aws-amplify/commit/257fc40))




<a name="0.4.3"></a>
## [0.4.3](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2...aws-amplify@0.4.3) (2018-06-02)


### Bug Fixes

* **aws-amplify:** Fix missing ServiceWorker export ([46069e2](https://github.com/aws/aws-amplify/commit/46069e2))
* **aws-amplify:** Fix missing ServiceWorker export ([acbf754](https://github.com/aws/aws-amplify/commit/acbf754))




<a name="0.4.2"></a>
## [0.4.2](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.30...aws-amplify@0.4.2) (2018-06-01)




**Note:** Version bump only for package aws-amplify
* **API:** query string for non signed requests ([#801](https://github.com/aws/aws-amplify/issues/801)) ([6746368](https://github.com/aws/aws-amplify/commit/6746368))




<a name="0.4.2-unstable.31"></a>
## [0.4.2-unstable.31](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.30...aws-amplify@0.4.2-unstable.31) (2018-06-02)


### Bug Fixes

* **aws-amplify:** Fix missing ServiceWorker export ([acbf754](https://github.com/aws/aws-amplify/commit/acbf754))




<a name="0.4.2-unstable.30"></a>
## [0.4.2-unstable.30](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.29...aws-amplify@0.4.2-unstable.30) (2018-06-01)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.29"></a>
## [0.4.2-unstable.29](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.28...aws-amplify@0.4.2-unstable.29) (2018-06-01)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.28"></a>
## [0.4.2-unstable.28](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.27...aws-amplify@0.4.2-unstable.28) (2018-06-01)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.27"></a>
## [0.4.2-unstable.27](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.26...aws-amplify@0.4.2-unstable.27) (2018-05-31)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.26"></a>
## [0.4.2-unstable.26](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.23...aws-amplify@0.4.2-unstable.26) (2018-05-31)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.23"></a>
## [0.4.2-unstable.23](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.22...aws-amplify@0.4.2-unstable.23) (2018-05-31)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.22"></a>
## [0.4.2-unstable.22](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.21...aws-amplify@0.4.2-unstable.22) (2018-05-31)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.21"></a>
## [0.4.2-unstable.21](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.20...aws-amplify@0.4.2-unstable.21) (2018-05-30)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.20"></a>
## [0.4.2-unstable.20](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.19...aws-amplify@0.4.2-unstable.20) (2018-05-29)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.19"></a>
## [0.4.2-unstable.19](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.18...aws-amplify@0.4.2-unstable.19) (2018-05-29)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.18"></a>
## [0.4.2-unstable.18](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.17...aws-amplify@0.4.2-unstable.18) (2018-05-29)


### Bug Fixes

* **Storage:** allowing to set an empty value as prefix for the S3 Key ([#931](https://github.com/aws/aws-amplify/issues/931)) ([240e50b](https://github.com/aws/aws-amplify/commit/240e50b))




<a name="0.4.2-unstable.17"></a>
## [0.4.2-unstable.17](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.16...aws-amplify@0.4.2-unstable.17) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.16"></a>
## [0.4.2-unstable.16](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.15...aws-amplify@0.4.2-unstable.16) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.15"></a>
## [0.4.2-unstable.15](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.14...aws-amplify@0.4.2-unstable.15) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.14"></a>
## [0.4.2-unstable.14](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.13...aws-amplify@0.4.2-unstable.14) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.13"></a>
## [0.4.2-unstable.13](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.12...aws-amplify@0.4.2-unstable.13) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.12"></a>
## [0.4.2-unstable.12](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.11...aws-amplify@0.4.2-unstable.12) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.11"></a>
## [0.4.2-unstable.11](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.10...aws-amplify@0.4.2-unstable.11) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.10"></a>
## [0.4.2-unstable.10](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.1...aws-amplify@0.4.2-unstable.10) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.9"></a>
## [0.4.2-unstable.9](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.8...aws-amplify@0.4.2-unstable.9) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.8"></a>
## [0.4.2-unstable.8](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.7...aws-amplify@0.4.2-unstable.8) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.7"></a>
## [0.4.2-unstable.7](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.6...aws-amplify@0.4.2-unstable.7) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.6"></a>
## [0.4.2-unstable.6](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.5...aws-amplify@0.4.2-unstable.6) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.5"></a>
## [0.4.2-unstable.5](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.4...aws-amplify@0.4.2-unstable.5) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.4"></a>
## [0.4.2-unstable.4](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.3...aws-amplify@0.4.2-unstable.4) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.3"></a>
## [0.4.2-unstable.3](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.2...aws-amplify@0.4.2-unstable.3) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.2"></a>
## [0.4.2-unstable.2](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.1...aws-amplify@0.4.2-unstable.2) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.1"></a>
## [0.4.2-unstable.1](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.2-unstable.0...aws-amplify@0.4.2-unstable.1) (2018-05-24)




**Note:** Version bump only for package aws-amplify

<a name="0.4.2-unstable.0"></a>
## [0.4.2-unstable.0](https://github.com/aws/aws-amplify/compare/aws-amplify@0.4.1...aws-amplify@0.4.2-unstable.0) (2018-05-23)




**Note:** Version bump only for package aws-amplify
