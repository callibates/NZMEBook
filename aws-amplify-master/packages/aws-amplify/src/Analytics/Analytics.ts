/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import {
    ConsoleLogger as Logger,
    missingConfig,
    Hub,
    Parser
} from '../Common';
import AWSPinpointProvider from './Providers/AWSPinpointProvider';
import Platform from '../Common/Platform';
import Auth from '../Auth';

import { AnalyticsProvider, EventAttributes, EventMetrics } from './types';

const logger = new Logger('AnalyticsClass');

const dispatchAnalyticsEvent = (event, data) => {
    Hub.dispatch('analytics', { event, data }, 'Analytics');
};

/**
* Provide mobile analytics client functions
*/
export default class AnalyticsClass {
    private _config;
    private _provider;
    private _pluggables: AnalyticsProvider[];
    private _disabled;
    private _autoSessionRecord;

    /**
     * Initialize Analtyics
     * @param config - Configuration of the Analytics
     */
    constructor() {
        this._config = {};
        this._pluggables = [];
        this._disabled = false;
    }

    /**
     * configure Analytics
     * @param {Object} config - Configuration of the Analytics
     */
    public configure(config?) {
        logger.debug('configure Analytics');
        if (!config) return this._config;

        const amplifyConfig = Parser.parseMobilehubConfig(config);
        this._config = Object.assign({}, this._config, amplifyConfig.Analytics, config);

        if (this._config['disabled']) {
            this._disabled = true;
        }

        if (this._config['autoSessionRecord'] === undefined) {
            this._config['autoSessionRecord'] = true;
        }

        this._pluggables.forEach((pluggable) => {
            // for backward compatibility
            if (pluggable.getProviderName() === 'AWSPinpoint' && !this._config['AWSPinpoint']) {
                pluggable.configure(this._config);  
            } else {
                pluggable.configure(this._config[pluggable.getProviderName()]);
            }
        });

        if (this._pluggables.length === 0) {
            this.addPluggable(new AWSPinpointProvider());
        }

        dispatchAnalyticsEvent('configured', null);
        logger.debug('current configuration', this._config);
        return this._config;
    }

    /**
     * add plugin into Analytics category
     * @param {Object} pluggable - an instance of the plugin
     */
    public addPluggable(pluggable: AnalyticsProvider) {
        if (pluggable && pluggable.getCategory() === 'Analytics') {
            this._pluggables.push(pluggable);
            let config = {};
            // for backward compatibility
            if (pluggable.getProviderName() === 'AWSPinpoint' && !this._config['AWSPinpoint']) {
                config = pluggable.configure(this._config);  
            } else {
                config = pluggable.configure(this._config[pluggable.getProviderName()]);
            }
            return config;
        }
    }

    /**
     * Get the plugin object
     * @param providerName - the name of the plugin 
     */
    public getPluggable(providerName) {
        for (let i = 0; i < this._pluggables.length; i += 1) {
            const pluggable = this._pluggables[i];
            if (pluggable.getProviderName() === providerName) {
                return pluggable;
            }
        }
      
        logger.debug('No plugin found with providerName', providerName);
        return null;
    }

    /**
     * Remove the plugin object
     * @param providerName - the name of the plugin
     */
    public removePluggable(providerName) {
        let idx = 0;
        while (idx < this._pluggables.length) {
            if (this._pluggables[idx].getProviderName() === providerName) {
                break;
            }
            idx += 1;
        }

        if (idx === this._pluggables.length) {
            logger.debug('No plugin found with providerName', providerName);
            return;
        } else {
            this._pluggables.splice(idx, idx + 1);
            return;
        }
    }

    /**
     * stop sending events
     */
    public disable() {
        this._disabled = true;
    }

    /**
     * start sending events
     */
    public enable() {
        this._disabled = false;
    }

    /**
    * Receive a capsule from Hub
    * @param {any} capsuak - The message from hub
    */
   public onHubCapsule(capsule: any): void {}


    /**
     * Record Session start
     * @return - A promise which resolves if buffer doesn't overflow
     */
    public async startSession(provider?: string) {
        const params = { event: { name: '_session_start' },  provider };
        return this._sendEvent(params);
    }

    /**
     * Record Session stop
     * @return - A promise which resolves if buffer doesn't overflow
     */
    public async stopSession(provider?: string) {
        const params = { event: { name: '_session_stop' }, provider };
        return this._sendEvent(params);
    }

    /**
     * Record one analytic event and send it to Pinpoint
     * @param {String} name - The name of the event
     * @param {Object} [attributs] - Attributes of the event
     * @param {Object} [metrics] - Event metrics
     * @return - A promise which resolves if buffer doesn't overflow
     */
    public async record(event: string | object, provider? , metrics?: EventMetrics) {
        let params = null;
        // this is just for compatibility, going to be deprecated
        if (typeof event === 'string') {
            params =  {
                'event': {
                    name: event, 
                    attributes: provider, 
                    metrics
                }, 
                provider: 'AWSPinpoint'
            };
        } else {
            params = { event, provider };
        }
        return this._sendEvent(params);
    }

    public async updateEndpoint(attrs, provider?) {
        const event = Object.assign({ name: '_update_endpoint' }, attrs);

        return this.record(event, provider);
    }

    private _sendEvent(params) {
        if (this._disabled) {
            logger.debug('Analytics has been disabled');
            return Promise.resolve();
        }

        const provider = params.provider? params.provider: 'AWSPinpoint';
        
        this._pluggables.forEach((pluggable) => {
            if (pluggable.getProviderName() === provider) {
                pluggable.record(params);
            }
        });

        return Promise.resolve();
    }
}
