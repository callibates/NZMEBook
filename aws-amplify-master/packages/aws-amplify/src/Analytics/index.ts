/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import AnalyticsClass from './Analytics';
import { AnalyticsProvider } from './types';

import {
    ConsoleLogger as Logger,
    Hub,
    Linking,
    AppState,
    Amplify
} from '../Common';
import Platform from '../Common/Platform';

const logger = new Logger('Analytics');
let startsessionRecorded = false;
let authConfigured = false;
let analyticsConfigured = false;

let _instance: AnalyticsClass = null;

if (!_instance) {
    logger.debug('Create Analytics Instance');
    _instance = new AnalyticsClass();
}

const Analytics = _instance;
Amplify.register(Analytics);

export default Analytics;
export { AnalyticsProvider };
export { AnalyticsClass };
export * from './Providers';

// listen on app state change
const dispatchAppStateEvent = (event, data) => {
    Hub.dispatch('appState', { event, data }, 'AppState');
};

if (Platform.isReactNative) {
    AppState.addEventListener('change', (nextAppState) => {
        switch(nextAppState) {
            case 'active':
                dispatchAppStateEvent('active', {});
        }
    });
}

// send a session start event if autoSessionRecord is enabled
const autoSessionRecord = () => {
    const config = Analytics.configure();
    startsessionRecorded = true;
    if (config.autoSessionRecord) {
        Analytics.updateEndpoint({}).then(() => {
            Analytics.startSession().catch(e => {
                logger.debug('start Session error', e);
            });
        });
    } else {
        logger.debug('auto Session record is diasabled');
    }            
};

Analytics.onHubCapsule = (capsule) => {
    const { channel, payload, source } = capsule;
    logger.debug('on hub capsule ' + channel, payload);

    switch(channel) {
        case 'auth':
            authEvent(payload);
            break;
        case 'storage':
            storageEvent(payload);
            break;
        case 'analytics':
            analyticsEvent(payload);
            break;
        default:
            break;
    }
};

const storageEvent = (payload) => {
    const { attrs, metrics } = payload;
    if (!attrs) return;

    Analytics.record('Storage', attrs, metrics);
};

const authEvent = (payload) => {
    const { event } = payload;
    if (!event) { return; }

    switch(event) {
        case 'signIn':
            Analytics.record('_userauth.sign_in');
            break;
        case 'signUp':
            Analytics.record('_userauth.sign_up');
            break;
        case 'signOut':
            break;
        case 'signIn_failure':
            Analytics.record('_userauth.auth_fail');
            break;
        case 'configured':
            authConfigured = true;
            if (authConfigured && analyticsConfigured && !startsessionRecorded) {
                autoSessionRecord();
            }
            break;
    }
};

const analyticsEvent = (payload) => {
    const { event } = payload;
    if (!event) return;

     switch(event) {
         case 'configured':
            analyticsConfigured = true;
            if (authConfigured && analyticsConfigured && !startsessionRecorded) {
                autoSessionRecord();
            }
            break;
     }
};

Hub.listen('auth', Analytics);
Hub.listen('storage', Analytics);
Hub.listen('analytics', Analytics);
