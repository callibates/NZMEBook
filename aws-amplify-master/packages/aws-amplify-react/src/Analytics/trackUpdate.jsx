/*
 * Copyright 2017-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance with
 * the License. A copy of the License is located at
 *
 *     http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
 * CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
 * and limitations under the License.
 */

import React, { Component } from 'react';
import { Analytics } from 'aws-amplify';

export function trackUpdate(Comp, trackerName) {
    return class extends Component {
        constructor(props) {
            super(props);
            this.trackerName = trackerName;
        }

        componentDidUpdate(prevProps, prevState) {
            const attributes = Object.assign({}, this.props, this.state);
            Analytics.record(this.trackerName, attributes);
        }

        render() {
            return <Comp {...this.props} />
        }
    }
}
