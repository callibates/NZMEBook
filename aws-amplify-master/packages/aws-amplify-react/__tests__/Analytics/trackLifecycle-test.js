import React, { Component } from 'react';
import { Analytics } from 'aws-amplify';
import { trackLifecycle } from '../../src/Analytics/trackLifecycle';

describe('trackLifecycle test', () => {
    test('render correctly', () => {
        const spyon = jest.spyOn(Analytics, 'record');
        const MockComp = class extends Component {
            render() {
                return <div />;
            }
        }
        const CompWithAuth = trackLifecycle(MockComp, 'trackername');
        const wrapper = shallow(<CompWithAuth/>);

        expect(wrapper).toMatchSnapshot();
        expect(spyon).toBeCalled();

        spyon.mockClear();
    });

    test('track when mounting', () => {
        const spyon = jest.spyOn(Analytics, 'record');
        const MockComp = class extends Component {
            render() {
                return <div />;
            }
        }
        const CompWithAuth = trackLifecycle(MockComp, 'trackername');
        const wrapper = mount(<CompWithAuth/>);
        
        expect(spyon.mock.calls.length).toBe(2);

        spyon.mockClear();
        wrapper.unmount();
    });

    test('track when unmounting', () => {
        const spyon = jest.spyOn(Analytics, 'record');
        const MockComp = class extends Component {
            render() {
                return <div />;
            }
        }
        const CompWithAuth = trackLifecycle(MockComp, 'trackername');
        const wrapper = shallow(<CompWithAuth/>);
        
        wrapper.unmount();
        expect(spyon.mock.calls.length).toBe(2);

        spyon.mockClear();
    });

    test('track when updating', () => {
        const spyon = jest.spyOn(Analytics, 'record');
        const MockComp = class extends Component {
            render() {
                return <div />;
            }
        }
        const CompWithAuth = trackLifecycle(MockComp, 'trackername');
        const wrapper = shallow(<CompWithAuth/>);

        wrapper.update();
        
        expect(spyon.mock.calls.length).toBe(2);

        spyon.mockClear();
    });
});