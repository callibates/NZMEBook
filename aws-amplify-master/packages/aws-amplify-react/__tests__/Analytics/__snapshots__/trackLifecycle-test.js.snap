// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`trackLifecycle test render correctly 1`] = `<MockComp />`;
